c10/cuda/CUDAGuard.h

```diff
@@ -76,7 +77,11 @@ struct CUDAGuard {
 /// CUDAGuard for when you can use this.
 struct OptionalCUDAGuard {
   /// Create an uninitialized OptionalCUDAGuard.
-  explicit OptionalCUDAGuard() : guard_() {}
+  explicit OptionalCUDAGuard() : guard_() {
+    // This constructor initializes CUDA context on current device.
+    // For more info see comment in CUDAFunctions.cpp
+    c10::cuda::SetTargetDevice();
```

Comment:
according to the comment in DeviceGuard.h, nullary constructor for OptionalCUDAGuard shouldn't initialize, current device at the time of construction doesn't matter at all. OptionalCUDAGuard should query device only at the time of initialization.  cc @ezyang

Reply:
Uninitialized OptionalDeviceGuard should literally do nothing (see the comment explaining its behavior), yet here you are unconditionally initializing context on current device, this can't be right?

Reply:
The current device must be set somehow along with `OptionalDeviceGuard` creation. I did not find a better place than these constructors. Otherwise, I would need to call `c10::cuda::SetTargetDevice()` after each Optional Guard is created. What would you recommend?

Reply:
No current device doesn't need to be set with default OptionalDeviceGuard construction, please see https://github.com/pytorch/pytorch/blob/ffd76d11c932138c924439f6fd74bb5efe78f69c/c10/core/DeviceGuard.h#L107-L118

Reply:
Natalia is absolutely right here. Is there a particular bug you were trying to fix with this?

Reply:
Okay, I reverted those changes in Optional Guards. Just trying to resolve test failures.